### PR TITLE
executor: fix CTE may be blocked when query report error

### LINKED ISSUE #31302: CTE is blocked when got error in Projection
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
go run [t.go](https://github.com/pingcap/tidb/files/7808267/t.txt) . Create table and insert rows.

    with recursive cte(iter, first, second, result) as (select 1, first, second, first+second from src  union all  select iter+1, second, result, second+result from cte where iter < 80 ) select * from cte;

### 2. What did you expect to see? (Required)

    mysql> with recursive cte(iter, first, second, result) as (select 1, first, second, first+second from src  union all  select iter+1, second, result, second+result from cte where iter < 80 ) select * from cte;
    ERROR 1690 (22003): BIGINT value is out of range in '(src.src.second + Column#14)'
### 3. What did you see instead (Required)
blocked
### 4. What is your TiDB version? (Required)

    +----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
    | tidb_version()
                                          |
    +----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
    | Release Version: v5.5.0-alpha-22-g68642193b9-dirty
    Edition: Community
    Git Commit Hash: 68642193b9d1105a2521009986efa10c9e85f1b7
    Git Branch: fix_cte_apply
    UTC Build Time: 2022-01-04 14:26:18
    GoVersion: go1.16.4
    Race Enabled: false
    TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
    Check Table Before Drop: false |
    +----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

### PR DESCRIPTION
**Pull request by guo-shaoge**
Signed-off-by: guo-shaoge <shaoge1994@163.com>

### What problem does this PR solve?

Issue Number: close #31302

Problem Summary:
```
mysql> explain with recursive cte(iter, first, second, result) as 
                       (select 1, first, second, first+second from src union all
                        select iter+1, second, result, second+result from cte where iter < 80 )
                        select * from cte;
+-------------------------------------+
| id                                  |
+-------------------------------------+
| CTEFullScan_18                      |
| CTE_0                               |
| ├─Projection_11(Seed Part)          |
| │ └─TableReader_13                  |
| │   └─TableFullScan_12              |
| └─Projection_14(Recursive Part)     |
|   └─Selection_15                    |
|     └─CTETable_16                   |
+-------------------------------------+
```
Why blocked:
 1. `Projection_14` will return overflow error
 2. ProjectionExec.Close() will be called, and it will wait for projectionInputFetcher to finish.
 3. projectionInputFetcher is blocked because CTETable_16 is blocked in `CTETable_16.Next()` because `CTETable_16` is waiting for e.iterInTbl.GetBegCh().
 4. `CTE_0` will reset iterInTbl.begCh when error occurs, so begCh is reset, but nobody close begCh anymore. So `CTETable_16` is blocked on `<- e.iterInTbl.GetBegCh()`

Why needs begCh:
 1. Some executors(like IndexLookUpJoin) start workers in `Open()` so it will call child.Next() in `Open()` phase.
 2. But data in iterInTbl is not ready in Open() phase, so these executor will read empty chunk which is unexpected.
 3. That's why we need begCh, it indicates when data is ready. When `computeSeedPart()` or each iteration in CTEExec is done, iterInTbl will be readable, we will close(begCh) and `CTETableExec` will continue to return correct chunk.

But it's too complicated and caused this block bug. So I decide to change code in `IndexLookUpJoin`  and remove `begCh`.

### What is changed and how it works?
For `IndexLookUpJoin`, `IndexNestedLookUpHashJoin` and `IndexLookUpMergeJoin`, start works in `Next()` instead of `Open()`.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [x] Performance regression: There maybe performance loss because workers in executor start in the first call of Next().
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
executor: fix CTE is block when query report error
```

### CODE REVIEW

File: executor/cte_test.go [hunk 1]
```diff
@@ -408,3 +410,33 @@ func TestSpillToDisk(t *testing.T) {
...
+	tk.MustExec("create table src(first int, second int);")
+
+	insertStr := fmt.Sprintf("insert into src values (%d, %d)", rand.Intn(1000), rand.Intn(1000))
+	for i := 0; i < 1000; i++ {
+		insertStr += fmt.Sprintf(",(%d, %d)", rand.Intn(1000), rand.Intn(1000))
+	}
+	insertStr += ";"
+	tk.MustExec(insertStr)
+
+	// Increase projection concurrency and decrese chunk size
```
Comment (guo-shaoge) [inline executor/cte_test.go:430]:
```suggestion
	// Increase projection concurrency and decrease chunk size
```

File: executor/index_lookup_join.go [hunk 2]
```diff
@@ -174,7 +175,7 @@ func (e *IndexLookUpJoin) Open(ctx context.Context) error {
 		e.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl.RegisterStats(e.id, e.stats)
 	}
 	e.cancelFunc = nil
-	e.startWorkers(ctx)
+	e.prepared = false
```
[the remark this answers is not in the document]
Comment (guo-shaoge) [reply executor/index_lookup_join.go]: done

**Review by XuHuaiyu — APPROVED** [review]
LGTM

**Pull request merged**